### DISCUSSION TITLE
Jvm target via Nashorn or Java or Clojure

### DISCUSSION THREAD

**Discussion by jkone27** [fable-compiler/Fable#3636] [Ideas]
would be cool and super duper if F# could target the JVM, which has kind of a big market share still, especially for existing backend services...

* [Nashorn](https://www.baeldung.com/java-nashorn) seems like JVM could run JS maybe...
* [JS 2 CLOJURE](https://roman01la.github.io/javascript-to-clojurescript/) or use Clojure as target lang
* [JS 2 Java](https://gist.github.com/thomd/5841814) some attempt to compile JS to Java
* [calva - clojure vscode extension repo](https://github.com/BetterThanTomorrow/calva)

**Comment by MangelMaxime** [comment]
The project you linked to seems to be transpilling JS to compile to Java/run over Java.

Because Fable generate JavaScript, you should be able to use the generated code as his.

If the goal is to support Java directly in Fable, then it means adding a new compiler target and for contributors to come to add it and maintain it.

**Comment by jkone27** [comment]
asking a bit in general if anyone knows, if one wanted to target Java trying to implement a test Fable.Java project, what would be the steps needed? 
* clojure target seems not awesome probably because is a lisp/based kind of dynamic lang , maybe targeting java is better for debugging? clojure is dynamic though and is an fp lang so should be easier.. [learn clojure in y minutes](https://learnxinyminutes.com/docs/clojure/)
* would ANTLR simplify the process? 
* could a small subset of F# be the beginning, for example.. just modules and functions for now (converted to static classes in java)

**Comment by jkone27** [comment]
other interesting facts:
 * clojure siding with a springboot project: https://www.linkedin.com/pulse/integrating-clojure-java-spring-boot-fernando-nunes-v2kaf/
 * https://linktosriram.github.io/clojure-spring-guide/
 * [groovy](https://learnxinyminutes.com/docs/groovy/) (also dynamic language but has better java integration and springboot support)
 * https://www.baeldung.com/groovy/spring-using-groovy
 * https://www.baeldung.com/spring-boot-groovy-web-app

**Comment by jkone27** [comment]
an old thread from F# compiler > https://github.com/fsharp/zarchive-fssf-ask-the-board/issues/4 , maybe relevant?

also chatgpt quick Q on the topic:

https://chatgpt.com/share/676430c0-e638-800e-b0ca-9e849516799f

**Comment by joprice** [comment]
Some other ml-style languages that compile to java might be of interest for research for java/bytecode generation:

* https://github.com/Frege/frege
* https://eta-lang.org/
* http://ocamljava.org/ <- seems to be unmaintained

**Comment by jkone27** [comment]
add also some libraries that might be useful for Java compilation world > AST / abstract synthax tree etc..

* [Eclipse JDT AST](https://www.vogella.com/tutorials/EclipseJDT/article.html)
* https://spoon.gforge.inria.fr/ | https://github.com/INRIA/spoon
* https://www.baeldung.com/java-poet (not mantained, has a more recent mantained fork but is in v 0.6
* https://www.antlr.org/ ? seems over complicated :)

**Comment by jkone27** [comment]
dropping this topic as probably is easier to "try" target Java as a regular Fable language if one wants to
